### PR TITLE
[CKW-5] Setup Back-end project

### PR DESCRIPTION
**Pull request by yeeezsh**
related to: [CKW-5](https://cpe-kmutt-web.atlassian.net/browse/CKW-5)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM node:14-slim AS dev
+RUN mkdir -p /app/backend
+WORKDIR /app/backend
+COPY package.json * /app/backend/
+COPY yarn.lock * /app/backend/
```
Comment (yeeezsh) [inline Dockerfile]: Why we need `*` it should simple like `COPY package.json/backend`

File: docker-compose.dev.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      DATABASE_CONNECTION: 'mongodb://mongodb-sharded:27017/cpe-kmutt-web'
+      DATABASE_USERAME: root
+      DATABASE_PASSWORD: CPE-KMUTT-DB
+      DATABASE_AUTH_SOURCE: admin
+      JWT_SECRET_KEY: very-secret-key.jwt
+      JWT_EXPIRES: 3600s
+      OPS_KEY: 'CPE-KMUTT.opskey'
+      NODE_ENV: 'development'
+    ports:
+      - 3000:3000
```
Comment (yeeezsh) [inline docker-compose.dev.yml:23]: same port should be `- 3000`

**Comment by kkerberoz** [comment]
Seen

**Comment by yeeezsh** [comment]
MongoDB cannot run correctly on `docker-compose.dev.yml`

File: backend/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+FROM node:14-slim AS dev
+RUN mkdir -p /backend
+WORKDIR /backend
```
Comment (yeeezsh) [inline backend/Dockerfile]: No need sub directory inside a docker container. It should be place on the `root`

File: backend/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+RUN yarn install
+COPY . /backend
+CMD ["yarn", "start"]
+EXPOSE 3000
+
+
+
+FROM node:14-slim AS build
+RUN mkdir -p /backend
+WORKDIR /backend
```
Comment (yeeezsh) [inline backend/Dockerfile]: same

File: backend/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "declaration": true,
+    "removeComments": true,
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
+    "allowSyntheticDefaultImports": true,
+    "target": "es2017",
+    "lib": ["ES2019", "ES2018"],
```
Comment (yeeezsh) [inline backend/tsconfig.json]: use the same name convention `ES2017`

File: docker-compose.dev.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      MONGODB_SHARDING_MODE: mongos
+      MONGODB_CFG_PRIMARY_HOST: mongodb-cfg
+      MONGODB_CFG_REPLICA_SET_NAME: cfgreplicaset
+      MONGODB_REPLICA_SET_KEY: CPE-KMUTT-DB-CLUSTER
+      MONGODB_ROOT_PASSWORD: CPE-KMUTT-DB
+      TZ: Asia/Bangkok
+    ports:
+      - "27017:27017"
+    volumes:
+      - "./mongo/mongos:/bitnami"
```
Comment (yeeezsh) [inline docker-compose.dev.yml]: cannot mount from root folder don't know why
use `local volume` instead

**Review by yeeezsh — COMMENTED** [review]
please add `. gitattributes` to set 
```
* text=auto eol=lf
```

File: backend/Dockerfile (hunk 3 above)
Comment (kkerberoz) [reply backend/Dockerfile]: Fixed

File: backend/Dockerfile (hunk 4 above)
Comment (kkerberoz) [reply backend/Dockerfile]: Fixed

File: backend/tsconfig.json (hunk 5 above)
Comment (kkerberoz) [reply backend/tsconfig.json]: Fixed

File: docker-compose.dev.yml (hunk 6 above)
Comment (nitpum) [reply docker-compose.dev.yml]: I suggest to use separate file `docker-compose.dev.win.yml` that contain only volume override 
and create script to run it
```bash
docker-compose -f docker-compose.yml -f docker-compose.dev.yml -f docker-compose.dev.win.yml up
```

Comment (nitpum) [reply docker-compose.dev.yml]: Example `docker-compose.dev.yml`
```yaml
services:
  mongos:
     volume:
       -  mongos:/bitnami

volumes:
  mongos:
```

File: backend/.dockerignore [hunk 7]
```diff
@@ -0,0 +1,13 @@
+.git
+.github
+.vscode
+build
+coverage
+docker-compose.yml
```
Comment (nitpum) [inline backend/.dockerignore:6]: ลืม `docker-compose.dev.yml` ครับ

**Review by nitpum — APPROVED** [review]
Anything else, LGTM

File: docker-compose.dev.yml [hunk 8]
```diff
@@ -23,12 +23,12 @@ services:
...
     links:
-      - mongos
-      - shard-0
+      - mongodb-sharded
+      - mongodb-shard0
       - mongodb-cfg
     restart: always
 
-  mongos:
+  mongodb-sharded:
```
Comment (yeeezsh) [inline docker-compose.dev.yml]: name should be `mongos`

Comment (kkerberoz) [reply docker-compose.dev.yml]: got this 
image

**Review by yeeezsh — COMMENTED** [review]
gogo

**Pull request merged**